### PR TITLE
Port time support to Rust

### PR DESCRIPTION
**Pull request by shaleh**

### CODE REVIEW

**Comment by shaleh** [comment]
This covers all of the easy bits.

Time arithmetic will be another PR.

The rest of the time functions require timezone handling. I plan to investigate the Chrono crate.

File: rust_src/src/time.rs [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+/// if successful, 0 if unsuccessful.
+#[no_mangle]
+pub extern "C" fn disassemble_lisp_time(
+    specified_time: Lisp_Object,
+    phigh: *mut Lisp_Object,
+    plow: *mut Lisp_Object,
+    pusec: *mut Lisp_Object,
+    ppsec: *mut Lisp_Object,
+) -> c_int {
+    let specified_time = LispObject::from(specified_time);
```
Comment (db48x) [inline rust_src/src/time.rs]: This is going to have to be `LispObject::from_raw` now.

File: rust_src/src/time.rs [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+    high: Lisp_Object,
+    low: Lisp_Object,
+    usec: Lisp_Object,
+    psec: Lisp_Object,
+    result: *mut lisp_time,
+    dresult: *mut f64,
+) -> c_int {
+    let high = LispObject::from(high);
+    let usec = LispObject::from(usec);
+    let psec = LispObject::from(psec);
```
Comment (db48x) [inline rust_src/src/time.rs]: Same with the `LispObject::from_raw`.

File: rust_src/src/time.rs [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+
+    if !(high.is_fixnum() && usec.is_fixnum() && psec.is_fixnum()) {
+        return 0;
+    }
+
+    let low = LispObject::from(low);
```
Comment (db48x) [inline rust_src/src/time.rs]: Same.

File: rust_src/src/time.rs [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+
+
+/// Return the current time, as the number of seconds since 1970-01-01 00:00:00.
+/// The time is returned as a list of integers (HIGH LOW USEC PSEC).
+/// HIGH has the most significant bits of the seconds, while LOW has the
+/// least significant 16 bits.  USEC and PSEC are the microsecond and
+/// picosecond counts.
+#[lisp_fn]
+pub fn current_time() -> LispObject {
+    LispObject::from(make_lisp_time(current_timespec()))
```
Comment (db48x) [inline rust_src/src/time.rs]: And again.

File: rust_src/src/time.rs [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+/// picosecond counts.
+#[lisp_fn]
+pub fn current_time() -> LispObject {
+    make_lisp_time_1(current_timespec())
+}
+
+/// Return the current time, as a float number of seconds since the
+/// epoch.  If TIME is given, it is the time to convert to float
+/// instead of the current time.  The argument should have the form
+/// \(HIGH LOW) or (HIGH LOW USEC) or (HIGH LOW USEC PSEC).  Thus, you
```
Comment (brotzeit) [inline rust_src/src/time.rs]: No need to escape in rust.

Comment (shaleh) [reply rust_src/src/time.rs]: Good catch.

**Comment by brotzeit** [comment]
The c code makes use of `lisp_h_XCONS` like #515.

**Comment by shaleh** [comment]
@brotzeit, yes it does. The Rust code usually has a better solution than asserts. Often our code is more explicit than the C code.

**Pull request merged**